contracts/pair_stable_bluna/src/contract.rs

```diff
@@ -0,0 +1,1810 @@
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::UpdateConfig { params } => update_config(deps, env, info, params),
+        ExecuteMsg::Receive(msg) => receive_cw20(deps, env, info, msg),
+        ExecuteMsg::ProvideLiquidity {
+            assets,
+            slippage_tolerance,
+            auto_stake,
+            receiver,
+        } => provide_liquidity(
+            deps,
+            env,
+            info,
+            assets,
+            slippage_tolerance,
+            auto_stake,
+            receiver,
+        ),
+        ExecuteMsg::Swap {
+            offer_asset,
+            belief_price,
+            max_spread,
+            to,
+        } => {
+            offer_asset.info.check(deps.api)?;
+            if !offer_asset.is_native_token() {
+                return Err(ContractError::Unauthorized {});
+            }
+
+            let to_addr = if let Some(to_addr) = to {
+                Some(addr_validate_to_lower(deps.api, &to_addr)?)
+            } else {
+                None
+            };
+
+            swap(
+                deps,
+                env,
+                info.clone(),
+                info.sender,
+                offer_asset,
+                belief_price,
+                max_spread,
+                to_addr,
+            )
+        }
+        ExecuteMsg::HandleReward {
+            previous_reward_balance,
+            old_user_share,
+            old_total_share,
+            user,
+        } => handle_rewards(
+            deps,
+            env,
+            info,
+            previous_reward_balance,
+            old_user_share,
+            old_total_share,
+            user,
+        ),
+    }
+}
+
+/// ## Description
+/// Receives a message of type [`Cw20ReceiveMsg`] and processes it depending on the received template.
+/// If the template is not found in the received message, then an [`ContractError`] is returned,
+/// otherwise returns the [`Response`] with the specified attributes if the operation was successful
+/// ## Params
+/// * **deps** is the object of type [`DepsMut`].
+///
+/// * **env** is the object of type [`Env`].
+///
+/// * **info** is the object of type [`MessageInfo`].
+///
+/// * **cw20_msg** is the object of type [`Cw20ReceiveMsg`].
+pub fn receive_cw20(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    cw20_msg: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let contract_addr = info.sender.clone();
+    match from_binary(&cw20_msg.msg) {
+        Ok(Cw20HookMsg::Swap {
+            belief_price,
+            max_spread,
+            to,
+        }) => {
+            // only asset contract can execute this message
+            let mut authorized: bool = false;
+            let config: Config = CONFIG.load(deps.storage)?;
+
+            for pool in config.pair_info.asset_infos {
```

Comment:
Maybe simpler, but not faster

Reply:
It is faster since `any` will stop execution when it recieves the first `true`. If you want the same behaviour then you need to add `break` after `authorized = true`

Reply:
Yes, it needs break operator to be faster. It was just copied code from pair_stable. There is a lot of refactoring we can do. Let's focus on functionality first

Reply:
Ok